### ISSUE TITLE
[java] LocalVariableCouldBeFinal: false positive with conditional variable assignment

### ISSUE THREAD

**Issue by StevanWhite** [pmd/pmd#6212]
**Affects PMD Version:** 7.17.0

**Rule:** [LocalVariableCouldBeFinal](https://docs.pmd-code.org/latest/pmd_rules_java_codestyle.html)

**Description:**

In a conditional variable assignment, PMD incorrectly states "Local variable 'a' could be declared final",
when simply specifying final on the variable would make the code fail to compile.

Granted, the code could be structured better.
This would be a perfect candidate for the ternary conditional operator, and with that, the variable could be declared final.
(Perhaps PMD could detect cases where the ternary conditional is appropriate.)

But the PMD complaint, taken literally, is incorrect.

**Code Sample demonstrating the issue:**

```java
/** PMD LocalVariableCouldBeFinal */
// pmd check -d=Final.java -R="pmd_rulesets.xml"
// gives
// LocalVariableCouldBeFinal:	Local variable 'a' could be declared final

public class
Final
{
	static int[][]
	ftn( boolean condition, int[][] b, int[][] c )
	{
		int[][] a; // local variable a could be final

		if( condition )
			a = b;
		else
			a = c;

		return a;
	}

	public static void
	main(String[] args)
	{
		final int b[][] = {{1}};
		final int c[][] = {{1},{2}};
		
		final int[][] q = ftn( false, b, c );

		System.out.println( q.length );
	}
}
```

**Steps to reproduce:**

Please provide detailed steps for how we can reproduce the bug.

1.  ... (e.g. if you're using maven: `mvn clean verify`)
2.  ...

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

**Comment by StevanWhite** [comment]
This is PMD 7.17.0

In a conditional variable assignment, PMD incorrectly states "Local variable 'a' could be declared final",
when simply specifying final on the variable would make the code fail to compile.

PMD does this in the attached example.

Granted, the code could be structured better.  
This would be a perfect candidate for the ternary conditional operator, and with that, the variable could be declared final.
(Perhaps PMD could detect cases where the ternary conditional is appropriate.)

But the PMD complaint, taken literally, is incorrect.

[Final.java.gz](https://github.com/user-attachments/files/23387655/Final.java.gz)

**Comment by adangel** [comment]
Please use the issue template the next time, thank you.

**Comment by adangel** [comment]
> when simply specifying final on the variable would make the code fail to compile.

I can't reproduce this. For me, the following code compiles:

```java
public class
Final
{
        static int[][]
        ftn( boolean condition, int[][] b, int[][] c )
        {
                final int[][] a; // local variable a could be final

                if( condition )
                        a = b;
                else
                        a = c;

                return a;
        }

        public static void
        main(String[] args)
        {
                final int b[][] = {{1}};
                final int c[][] = {{1},{2}};

                final int[][] q = ftn( false, b, c );

                System.out.println( q.length );
        }
}
```

**Comment by adangel** [comment]
From my perspective, everything works, the rule is correct. `a` can be made final.

**Comment by StevanWhite** [comment]
I can't imagine why I didn't try my example.  I pulled out of more complex code, where it wasn't so clear.

I guess I was just so sure that a final couldn't be assigned.... but then again, final member variables are initialized in constructors.

My apologies. Closing.
